### DISCUSSION TITLE
async defaultValues in React Native?

### DISCUSSION THREAD

**Discussion by GollyJer** [react-hook-form/react-hook-form#4139] [Q&A] [answered]
Hi! I've been banging my head on this one for a while. It's the only thing keeping us from switching over from Formik.

I searched and found plenty of discussion on this topic and tried to implement [this solution](https://github.com/react-hook-form/react-hook-form/discussions/2282).

It looks like this.
```javascript
const me = useMe(); // 👈 This is an initially empty graphql call that renders again when it has data.

const defaultValues = {
  first_name: me?.first_name || '',
};

const { control, errors, reset, handleSubmit } = useForm({
  defaultValues,
});

useEffect(() => reset(defaultValues), [reset]);
```
and later on in the component render...
```javascript 
<Controller
  name="first_name"
  control={control}
  render={({ onChange, value }) => (
    <TextInput
      label="First Name"
      textColor={theme.colors.textInput.color}
      baseColor={theme.colors.textInput.color}
      value={value}
      error={errors.first_name}
      onChangeText={(v) => onChange(v)}
    />
  )}
/>
```

However, it seems defaultValues is never updated even the `me` comes back with the data.

What am I doing wrong? Thanks!

**Comment by bluebill1049** [comment]
hey @GollyJer if `me` is coming async, then you will need to use `setValue` or `reset`.

`useEffect(() => reset(defaultValues), [reset]);` are you missing defaultValue in the dep array?

**Reply by GollyJer** [reply]
Thanks for the super fast reply @bluebill1049. 

If I add `defaultValues`, or `me`, to the dependency array it creates a render loop.

**Reply by bluebill1049** [reply]
you probably need to `useMemo`. because without `defaultValues`, it's always a new reference on each render.

you can also consider passing `defaultValueas` as props and then reset with useEffect.

**Reply by GollyJer** [reply]
Good catch. I updated defaultValues to...

```javascript 
  const defaultValues = useMemo(
    () => ({
      first_name: me?.first_name || '',
    }),
    [me?.first_name]
  );
```
and the useEffect to this...
```javascript 
useEffect(() => {
    console.log(defaultValues);
    reset(defaultValues);
  }, [defaultValues, reset]);
```

It works exactly as expected. The console log prints.
`{ first_name: '' }`
`{ first_name: 'Jeremy' }`

And the form renders one last time after `reset(defaultValues)` is called but the `Controller` still doesn't update the value in the form. 😿

**Reply by bluebill1049** [reply]
have a play here: https://codesandbox.io/s/lingering-moon-0kost?file=/src/App.js do a console.log at your controller see if props actually pass down.

**Reply by GollyJer** [reply]
Thanks @bluebill1049. That helped.

The `TextInput` itself doesn't render when it gets a new value. I was able to get around it by not rendering until the value shows up.

Like this...
```javascript 
<Controller
  name="first_name"
  control={control}
  render={({ onChange, value }) => (
    <>
      {!!value && (
        <TextInput
          label="First Name"
          textColor={theme.colors.textInput.color}
          baseColor={theme.colors.textInput.color}
          value={value}
          error={errors.first_name}
          onChangeText={(text) => {
            onChange(text);
          }}
        />
      )}
    </>
  )}
/>
```

**Comment by GollyJer** [comment, marked as answer]
The full solution to mark as the answer.

```javascript 
const me = useMe(); // 👈 This is an initially empty graphql call that renders again when it has data.

const defaultValues = useMemo(
  () => ({
    first_name: me?.first_name || '',
  }),
  [me?.first_name]
);

const { control, errors, reset, handleSubmit } = useForm({
  defaultValues,
});

useEffect(() => {
  reset(defaultValues);
}, [defaultValues, reset]);
```

and later on in the render...
```javascript 
<Controller
  name="first_name"
  control={control}
  render={({ onChange, value }) => (
    <>
      {!!value && (
        <TextInput
          label="First Name"
          textColor={theme.colors.textInput.color}
          baseColor={theme.colors.textInput.color}
          value={value}
          error={errors.first_name}
          onChangeText={(text) => {
            onChange(text);
          }}
        />
      )}
    </>
  )}
/>
```

**Reply by bluebill1049** [reply]
oh! no, I bet you are missing a `defaultValue` or `defaultValues` at useForm

**Reply by sebastienbarre** [reply]
Thanks, that `useEffect()` solution is very helpful. @GollyJer just out of curiosity, why the `useMemo()` before it?

**Reply by nani1008** [reply]
@sebastienbarre I think it will update the state if you wont cache the values using useMemo()

**Reply by GollyJer** [reply]
> Thanks, that `useEffect()` solution is very helpful. @GollyJer just out of curiosity, why the `useMemo()` before it?

https://github.com/react-hook-form/react-hook-form/discussions/4139#discussioncomment-351405
